### PR TITLE
 Running connector framework on top of functions 

### PR DESCRIPTION
**Pull request by jerrypeng**
### Motivation

Pulsar connectors framework running on top of pulsar functions

### CODE REVIEW

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+        void processArguments() throws Exception {
+        }
+
+        abstract void runCmd() throws Exception;
+    }
+
+    @Parameters(commandDescription = "Create Pulsar source connectors")
+    class CreateSource extends BaseCommand {
+        @Parameter(names = "--fqfn", description = "The Fully Qualified Function Name (FQFN) for the function")
```
Comment (srkukarni) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java]: Can we remove this and instead assume for now that all connectors are running in public/default tenant/namespace? And then based on how Oath wants to run it revisit this question later. IMO this exposes too much the function underpinnings of the connect framework

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        protected String fqfn;
+        @Parameter(names = "--tenant", description = "The function's tenant")
+        protected String tenant;
+        @Parameter(names = "--namespace", description = "The function's namespace")
+        protected String namespace;
+        @Parameter(names = "--name", description = "The function's name")
+        protected String name;
+        @Parameter(names = "--className", description = "The function's class name")
+        protected String className;
+        @Parameter(names = "--topicName", description = "Pulsar topic to ingress data to")
```
Comment (srkukarni) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java]: Maybe we should have inputTopicName and outputTopicName instead?

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+        void runCmd() throws Exception {
+            if (!areAllRequiredFieldsPresentForSource(sourceConfig)) {
+                throw new RuntimeException("Missing arguments");
+            }
+            admin.functions().createFunction(convertSourceConfig(sourceConfig), jarFile);
+            print("Created successfully");
+        }
+    }
+
+    @Parameters(commandDescription = "Create Pulsar sink connectors")
```
Comment (srkukarni) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java]: CreateSink and CreateSource look almost the same. Maybe refactoring?

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+                names = "--jar",
+                description = "Path to the jar file for the function (if the function is written in Java)",
+                listConverter = StringConverter.class)
+        protected String jarFile;
+
+        @Parameter(names = "--functionConfigFile", description = "The path to a YAML config file specifying the "
+                + "function's configuration")
+        protected String fnConfigFile;
+
+        protected SourceConfig sourceConfig;
```
Comment (srkukarni) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdConnectors.java]: maybe unify SourceConfig/SinkConfig to ConnectConfig?

File: pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/PushSource.java [hunk 5]
```diff
@@ -48,5 +48,5 @@
      * to pass messages whenever there is data to be pushed to Pulsar.
      * @param consumer
      */
-    void setConsumer(Function<T, CompletableFuture<Void>> consumer);
-}
\ No newline at end of file
+    void setConsumer(Function<T, CompletableFuture<Void>> consumer) throws Exception;
```
Comment (srkukarni) [inline pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/PushSource.java]: why do we need this change?

Comment (jerrypeng) [reply pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/PushSource.java]: if the user wants to throw checked exception but not wanting to wrap it around a try catch and throw runtimeException.  Its kinda unnecessary

File: pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Sink.java [hunk 6]
```diff
@@ -45,5 +45,5 @@
      * @param message Object to publish to the sink
      * @return Completable future fo async publish request
      */
-    CompletableFuture<Void> write(final T message);
-}
\ No newline at end of file
+    CompletableFuture<Void> write(final T message) throws Exception;
```
[the remark this answers is not in the document]
Comment (jerrypeng) [reply pulsar-connect/core/src/main/java/org/apache/pulsar/connect/core/Sink.java]: if the user wants to throw checked exception but not wanting to wrap it around a try catch and throw runtimeException.  Its kinda unnecessary

**Pull request closed without merging**